### ISSUE TITLE
okta-react LoginCallback not displaying error if user not assigned application

### ISSUE THREAD

**Issue by swiftone** [okta/okta-oidc-js#926]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [x] okta-react
- [ ] okta-react-native

## I'm submitting a:

- [x] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior
(Note - this is copied from https://github.com/okta/okta-oidc-js/issues/719#issuecomment-705216112 )

-  In the case of: OAuthError (eg. "User is not assigned to the client application.")
- the LoginCallback component does not display the error message 
  - Shows a blank page on /implicit/callback
- A custom errorComponent does not work either

## Expected behavior
- Show error message

## Minimal reproduction of the problem with instructions
- Login with a user that isn't assigned to the client application
- After a user logs in, the LoginCallback component does not display the error message

## Extra information about the use case/user story you are trying to implement

Moving this to a new issue as this is a new investigation

## Environment

- Package Version: 3.0.4, previously this was reported fixed as of 3.0.1

**Comment by swiftone** [comment]
Created this issue off of https://github.com/okta/okta-oidc-js/issues/719#issuecomment-705216112 to investigate.

**Comment by swiftone** [comment]
@CristianoYL - I'll take a stab at recreating in the near future.  
- Can you give an example of how you attempted to use errorComponent?
- Do you see the OAuth error in the network traffic?

**Comment by CristianoYL** [comment]
@swiftone Thanks for your quick response. I can reproduce the error using the minimal setup as in the tutorial, and `errorComponent` seems to be irrelevant to this issue (explained later).

I took a look at the source code and noticed that #848 reverts your changes in #738, which was meant to maintain the error in `authState`. 

When I log `authState` to console in the **LoginCallback** component, I noticed that the error gets overwritten. (case 1, using latest version)
![image](https://user-images.githubusercontent.com/17320996/95397365-e2150d80-08b7-11eb-969d-e563e9e9b86c.png)

If I revert the changes, the error can be persisted so the component renders it as expected. However, if the user can indeed authenticate, the app will not redirect (case 2). My naive guess is that the definition of `authState.isPending` has changed over time, since I noticed it's always true in case 2 after coming back from the Okta-hosted login page.
![image](https://user-images.githubusercontent.com/17320996/95397875-32409f80-08b9-11eb-8b01-04c908b72140.png)
So `authService.handleAuthentication()` can never be executed in the effect.

My temporary workaround is to register the latest non-null error as state:
```
    const { authService, authState } = useOktaAuth();
    const [error, setError] = useState(null);

    useEffect(() => {
        if (authState.error) {
            setError(authState.error);
        }
        if (authState.isPending) {
            authService.handleAuthentication();
        }
    }, [authService, authState.isPending, authState.error]);

    if (!authState.isPending && error) {
        // ... render error component
```

**Comment by kellengreen** [comment]
I can reproduce in `4.0.1`.

**Comment by aarongranick-okta** [comment]
@kellengreen We are investigating this issue. It may be related to an issue which was recently fixed in `okta-auth-js`. Deleting your `node_modules` and reinstalling *may* fix the issue.  I will work to reproduce the issue. Internal ref: OKTA-361433

**Comment by aarongranick-okta** [comment]
@kellengreen I think this issue was fixed in `okta-auth-js@4.6.0`. Current versions of `okta-react` allow you to pass your own instance of `OktaAuth`. Upgrading this instance to `4.6.0` should fix the issue you are seeing.

**Comment by kellengreen** [comment]
We're on `4.6.0` currently.

**Comment by aarongranick-okta** [comment]
@kellengreen I've been able to reproduce the issue and see the problem. The `LoginCallback` in `okta-react` is only showing errors from the `authState` object, however in this case the error is being thrown from `handleLoginRedirect`, which is called by `LoginCallback`. `LoginCallback` does not catch the error here, as is done in `okta-angular`: https://github.com/okta/okta-angular/blob/master/src/okta/components/callback.component.ts#L28


I've created a new internal issue OKTA-361608 to fix `LoginCallback` component to catch this error and display it. Hopefully our team can get this fixed soon.

As a workaround, in the meantime you could provide your own callback component. The only thing it needs to do is call `handleLoginRedirect` exactly once (and catch/display any thrown error).  

If you do come up with a great fix for the `LoginCallback` we would be happy to accept it as a PR.

**Comment by kellengreen** [comment]
Ooh good I'm glad you were able to reproduce.

> As a workaround, in the meantime you could provide your own callback component.

Thanks, that's what I ended up doing in the interim.

```javascript 
export default function CustomLoginCallback() {
  const query = useQuery()
  return query.error_description ? (
    <Alert severity="error">
      <AlertTitle>Authentication Failure</AlertTitle>
      {query.error_description}
    </Alert>
  ) : (
    <LoginCallback />
  )
}
```
Sidenote / random thought, it would be nice to be able to customize error result of `<LoginCallback/>`. For example: 

```
<LoginCallback 
  errorComponent={({ error }) => (
    <Alert severity="error">
      <AlertTitle>Authentication Failure</AlertTitle>
        {error.message}
    </Alert>
  )}
/>
```

**Comment by aarongranick-okta** [comment]
@kellengreen It is not well documented but the `LoginCallback` does accept an `errorComponent` prop: https://github.com/okta/okta-react/blob/master/src/LoginCallback.tsx#L23

**Comment by kellengreen** [comment]
Aah [so it does](https://www.npmjs.com/package/@okta/okta-react#general-components) 👍

**Comment by aarongranick-okta** [comment]
@CristianoYL @kellengreen We have released a fix for this issue. After upgrading to `@okta/okta-react` [version 5.1.1](https://github.com/okta/okta-react/releases/tag/okta-react-5.1.1) you should see OAuth callback errors caught and displayed in the `LoginCallback` component.

**Comment by lyin-vir** [comment]
@aarongranick-okta I can confirm the issue is fixed on v5.1.1. Thanks for the update.

I noticed that the documentation is a little lagged behind. For anyone who wants to use the custom error component, here's what I did:

``` jsx
import React from 'react';
import { LoginCallback } from '@okta/okta-react';

const CustomLoginCallback = () => {

    const renderError = (props) => {
        const { error } = props;
        if (error) {
            const { name, errorCode, errorSummary } = error;

            if (errorCode === 'access_denied') {  # customized error handling
                return (
                    <div>
                        <p>You haven't been assigned to this app yet.</p>
                    </div>
                );
            }

            # genric error handling
            return (
                <div>
                  <p>{[${name}: ${errorSummary}]`}</p>
                </div>
            );
        }

        return null;
    };
    return <LoginCallback errorComponent={renderError} />;
};

export default CustomLoginCallback;
```

Then in your App, or wherever you defined your routes, use this `CustomLoginCallback` instead of `LoginCallback`.

**Comment by CristianoYL** [comment]
@aarongranick-okta I can confirm the issue is fixed on v5.1.1. Thanks for the update.

I noticed that the documentation is a little lagged behind. For anyone who wants to use the custom error component, here's what I did:

``` jsx
import React from 'react';
import { LoginCallback } from '@okta/okta-react';

const CustomLoginCallback = () => {

    const renderError = (props) => {
        const { error } = props;
        if (error) {
            const { name, errorCode, errorSummary } = error;

            if (errorCode === 'access_denied') {  # customized error handling
                return (
                    <div>
                        <p>You haven't been assigned to this app yet.</p>
                    </div>
                );
            }

            # genric error handling
            return (
                <div>
                  <p>{[${name}: ${errorSummary}]`}</p>
                </div>
            );
        }

        return null;
    };
    return <LoginCallback errorComponent={renderError} />;
};

export default CustomLoginCallback;
```

Then in your App, or wherever you defined your routes, use this `CustomLoginCallback` instead of `LoginCallback`.

**Comment by sscho** [comment]
Thanks all. Here's what I did:

`<Route path="/login/callback" component={() => <LoginCallback errorComponent={LoginCallbackError} />} />`

Where LoginCallbackError is my custom callback component. You can access props.error to find the error message.

**Comment by sdtacoma** [comment]
@CristianoYL 

Thank you for this, it is exactly what I needed. I believe the code snippet is missing a back-tick in the 'general error handling' return.

```
<p>{`[${name}: ${errorSummary}]`}</p>
```

**Comment by denysoblohin-okta** [comment]
Closing this issue as it was fixed in 5.1.1
